### DISCUSSION TITLE
OpenScope changes

### DISCUSSION THREAD

**Discussion by timrhaynes** [dadhi/DryIoc#401] [Q&A]
In DryIoc 2.12.10, OpenScope returns an IContainer, but in DryIoc 4.7.5, OpenScope returns an IResolverContext.
How would I rewrite this method to work properly under 4.7.5?
Thank you!

```
  public IContainer OpenScope(string scopeName)
        {
            var container = _internalDryContainer.OpenScope(scopeName);

            container.Register<IScopedDIContainer, DIContainer>(Reuse.InCurrentScope, Made.Of(FactoryMethod.ConstructorWithResolvableArgumentsIncludingNonPublic), ifAlreadyRegistered: IfAlreadyRegistered.Replace);
            return container.Resolve<IScopedDIContainer>();
        }
        
 
```

**Comment by dadhi** [comment]
@timrhaynes The direct rewrite would be something like this:

```cs
 public IContainer OpenScope(string scopeName)
 {
     var container = _internalDryContainer.OpenScope(scopeName).Resolve<IContainer>(); // see the Resolve here!
     container.Register<IScopedDIContainer, DIContainer>(Reuse.InCurrentScope,
         Made.Of(FactoryMethod.ConstructorWithResolvableArgumentsIncludingNonPublic), ifAlreadyRegistered: IfAlreadyRegistered.Replace);
      return container.Resolve<IScopedDIContainer>();
}
```

But additionally I would use the `Use` instead of  `container.Register<IScopedDIContainer, DIContainer>(...)` because the `Use` is specifically designed to add the runtime state into the open scope:
```cs
 public IContainer OpenScope(string scopeName)
 {
     var scope = _internalDryContainer.OpenScope(scopeName);
     var diContainer = new DiContainer();
     scope.Use<IScopedDIContainer>(new DiContainer());
     return diContainer;
}
```

**Reply by timrhaynes** [reply]
Perfect, thank you! I'm a bit lost on the .Use() method, and cannot seem to locate the docs for it. To me, this reads kind of like:
```
 var A = new A();
 var B = new B();
 a.Something();
 return B;
```
 
 But... B wasn't really used at all, so why are we returning it? Are there docs for the .Use() method that can clear this up for me? Thanks again!

**Reply by dadhi** [reply]
@timrhaynes Check the 2nd example here https://github.com/dadhi/DryIoc/blob/master/docs/DryIoc.Docs/RegisterResolve.md#registerinstance

I will probably make it into sub-topic.

**Reply by timrhaynes** [reply]
> @timrhaynes The direct rewrite would be something like this:
> 
[removed for brevity, just look up above]

The initial code has different container state after executing, than your new code does:

When debugging the initial code, with a breakpoint at the end, and then checking for a certain controller's registration in both the regular and scoped containers with the VS2019 Immediate window:

_internalDryContainer.IsRegistered<My.Controllers.HomeController>()
true
container.Resolve<IScopedDIContainer>().IsRegistered<My.Controllers.HomeController>()
true

When debugging the new "Use" code, with a breakpoint at the end, and then checking for a certain controller's registration in both the regular and scoped containers in the VS2019 Immediate window:
_internalDryContainer.IsRegistered<Trader.Marketplace.Responsive.Web.Controllers.HomeController>()
true
diContainer.IsRegistered<Trader.Marketplace.Responsive.Web.Controllers.HomeController>()
**false**

Maybe there is a bug in your new code?

**Reply by dadhi** [reply]
I am not sure about the full context of your code and why are you checking with `IsRegistered` but the `Use` is not a real registration, so can't be checked by `IsRegistered`. `Use` supposed to directly add some data to scope. If you need the real registration, then you may try to exploit the Child Container for that.

But first I would suggest to go over your setup and ask why do you need this complexity. To help with that imagine yourself solving the task in question without DI container, by calling constructors and methods, and manipulating the state. Then automate you manual setup with DI container.

**Reply by timrhaynes** [reply]
Okay, I will give that some more thought, thank you!
